subsys/zbus/zbus.c

```diff
@@ -191,21 +194,133 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
 	return last_error;
 }
 
+static inline void _zbus_chan_highest_priority_observer_update(const struct zbus_observer *obs)
+{
+	struct zbus_channel_observation *observation;
+
+	int count;
+
+	STRUCT_SECTION_COUNT(zbus_channel_observation, &count);
+
+	atomic_val_t obs_prio_val;
+	atomic_val_t chan_high_obs_prio_val;
+
+	for (int16_t i = 0; i < count; ++i) {
+		STRUCT_SECTION_GET(zbus_channel_observation, i, &observation);
+
+		__ASSERT(observation != NULL, "observation must be not NULL");
+
+		const struct zbus_channel *chan = observation->chan;
+
+		if (obs != observation->obs) {
+			continue;
+		}
+
+		obs_prio_val = atomic_get(&obs->data->priority);
+		chan_high_obs_prio_val = atomic_get(&chan->data->highest_observer_priority);
+
+		if ((obs_prio_val > 0) && (chan_high_obs_prio_val > obs_prio_val)) {
+			atomic_cas(&chan->data->highest_observer_priority, chan_high_obs_prio_val,
+				   obs_prio_val);
```

Comment:
I'm a little confused at this usage: atomic_cas() can fail if it races on the old value, and this isn't handling that case?  Is that because it can't fail by construction?  If so, you'd probably be better off using regular locking and normal variables instead of atomics.  If not, this needs to be fixed.

Reply:
This still needs a fix.  To clarify from earlier: atomic_cas() will swap the two priority values if and only if the old value still stored in the atomic variable has not changed from "obs_prio_val".  If it has, then atomic_cas() returns a failure condition that needs to be handled (usually with a retry of the get/compute/set sequence) by the caller.

The only way this isn't a bug is if it's impossible by construction that two threads can race on the highest_observer_priority field.  But I don't see any locking anywhere that would prevent that.